lib/src/common/managed_stream_subscription.dart

```diff
@@ -0,0 +1,42 @@
+import 'dart:async';
+
+/// An implementation of StreamSubscription that provides a didCancel future.
+class ManagedStreamSubscription<T> extends StreamSubscription<T> {
+  final StreamSubscription<T> _subscription;
+
+  Completer<Null> _didCancel = new Completer();
+
+  ManagedStreamSubscription(Stream<T> stream, void onData(T))
+      : _subscription = stream.listen(onData);
+
+  @override
+  Future<E> asFuture<E>([E futureValue]) => _subscription.asFuture(futureValue);
+
+  @override
+  Future<Null> cancel() async {
```

Comment:
Let's not make this async since it isn't on the base class. Maybe just chain the completion onto the end of the subscription's cancel future. Although that can be null, so you'll want to null check it and build a throwaway future (maybe) otherwise.